src/ds_flood_gfm/datasources/gfm.py

```diff
@@ -182,9 +385,147 @@ def export_polygons(gdf: gpd.GeoDataFrame, output_path: Path, local=False, blob=
+    """Create provenance raster showing last observation date per pixel.
+
+    Parameters
+    ----------
+    stack_flood_max : xr.DataArray
+        3D array (time, y, x) of flood data.
+    unique_dates : np.ndarray
+        Array of datetime64 dates corresponding to time dimension.
+
+    Returns
+    -------
+    tuple[xr.DataArray, dict]
+        - provenance_idx: 2D lazy DataArray with integer indices (NOT computed)
+        - date_mapping: Dictionary mapping indices to date strings
+    """
+    logger.info("Creating provenance raster (lazy)...")
+
+    # Create mask of where we have valid observations (regardless of flood value)
+    has_data = ~stack_flood_max.isnull()
+
+    # Find the LAST time index where we had data for each pixel
+    # argmax finds first True, so we reverse the time dimension first
+    # This stays LAZY until .compute() is called
+    has_data_reversed = has_data.isel(time=slice(None, None, -1))
+    provenance_idx_reversed = has_data_reversed.argmax(dim='time', skipna=True)
+
+    # Convert back to original time indexing
+    n_times = len(unique_dates)
+    provenance_idx = n_times - 1 - provenance_idx_reversed
+
+    # Mask pixels that never had data
+    never_had_data = ~has_data.any(dim='time')
+    provenance_idx = provenance_idx.where(~never_had_data, -1)
+
+    # Create date mapping for lookup
+    date_mapping = {
+        int(i): str(pd.Timestamp(date))[:10]
+        for i, date in enumerate(unique_dates)
+    }
+    date_mapping[-1] = "No Data"
+
+    logger.info(f"  Provenance raster created (lazy, shape will be {provenance_idx.shape})")
+    logger.info(f"  Date mapping: {date_mapping}")
+
+    return provenance_idx, date_mapping
+
+
+def add_modal_provenance_to_admin(gdf_admin: gpd.GeoDataFrame,
+                                   provenance_idx: xr.DataArray,
+                                   date_mapping: dict,
+                                   admin_level: int = 3) -> gpd.GeoDataFrame:
+    """Add modal (most common) provenance date to admin boundaries.
+
+    Uses exactextract to efficiently calculate the most common provenance date
+    within each administrative boundary.
+
+    Parameters
+    ----------
+    gdf_admin : gpd.GeoDataFrame
+        Administrative boundaries.
+    provenance_idx : xr.DataArray
+        Provenance raster with integer indices (should be computed, not lazy).
+    date_mapping : dict
+        Mapping from integer indices to date strings.
+    admin_level : int, default 0
+        Administrative level (0, 1, 2, 3) for column naming.
+
+    Returns
+    -------
+    gpd.GeoDataFrame
+        Admin boundaries with 'prov_date' and 'prov_idx' columns added.
+    """
+    logger.info(f"Adding modal provenance to admin level {admin_level} boundaries...")
+
+    # Write provenance raster to temporary file for exactextract
+    with tempfile.NamedTemporaryFile(suffix=".tif", delete=False) as tmp:
```

Comment:
Not sure if this is more performant, but you should also be able to pass the raster in memory (eg. as an xr.DataArray object) to `exactextract`.